rfcs/2020-11-02-remap-infallible-functions.md

```diff
@@ -0,0 +1,371 @@
+
+While the idea is that this is the most simple solution to the problem, there is
+a case to be made for the fact that this can unintentionally silently ignore
+errors that should be caught, if one forgets to add the `!` character, or simply
+isn't aware of the existence of fallible and infallible function variants.
+
+It is therefor possible to decide that we want the inverse to happen, without a
+bang the function fails at runtime, with a bang, you accept the suppression of
+errors (and potentially manually handle those errors).
+
+### Manual Abort
+
+If a user wants to manually abort the program for a given failure, they
+currently can't.
+
+Take this example:
+
+```javascript
+if ($message = parse_json(.message); is_undefined($message)) {
+    // parsing failed
+
+    if .foo > 5 {
+        .message = "failed, but foo was greater than 5, so all's good!"
+    } else {
+        // abort, abort!
+    }
+} else {
+    // parsing succeeded
+}
+```
+
+We could add an `abort` or `fail` function which takes a string, and returns an
+error containing the given string.
+
+This would allow one to write:
+
+```rust
+if /* condition */ {
+    // success
+} else {
+    abort("failed, because foo was less than 5, we really can't let this slide!")
+}
+```
+
+This change is currently **out of scope** for this proposal.
+
+## Doc-level Proposal
+
+- If a user wants event processing to fail if a function fails, they add a bang
+  (`!`) to the function identifier (before the parenthesis and its arguments).
+- They can manually detect if a function fails through if statements and
+  `is_undefined`.
+
+## Rationale
+
+Remap relies heavily on functions to map object data. A lot of these functions
+can fail, and if one fails, all subsequent operations on the object are aborted.
+
+This can result in unexpected outcomes, which we want to avoid as much as
+possible.
+
+The proposed change is fairly limited in scope, it adds a bit of (additive)
+syntax change to the language, only a minor refactor to the parser, and a simple
+implementation of the new `Infallible` expression.
+
+All in all, the changes required are minimal and non-invasive, while the upside
+is more control for the end-user on when to fail program execution and when not
+to.
+
+## Prior Art (TODO)
+
+## Drawbacks (TODO)
+
+- added complexity to the language
+- "safe by default" might obscure errors
+
+## Alternatives (TODO)
+
+- do nothing
+- try blocks
+- explicit unwrapping of function results
+
+## Outstanding Questions
+
+- Should we use "safe by default" or "fallible by default"?
+- Should assigning an undefined value result in no action, or removing any
+  potential existing value of the target path/variable?
+- How should if statements and logical operators handle undefined values?
+- Should we provide compile-time error for fallibly calling always-infallible
+  functions (e.g. `now!()`)?
+- Should we add support for (in)fallible path queries in this RFC?
```

Comment:
I'd defer this and the manual abort function. They seem like they can be handled separate, and after, this RFC.